common/achievement/get.ts

```diff
@@ -0,0 +1,99 @@
+import { prisma } from '../prisma';
+import { Achievement_template, User_achievement, achievement_action_type_enum, achievement_type_enum } from '../../graphql/generated';
+import { Step, Achievement, achievement_state } from '../../graphql/types/achievement';
+import { User } from '../user';
+import { renderTemplate } from '../../utils/helpers';
+import { Context } from '../notification/types';
+
+const getUserAchievements = async (user: User): Promise<Achievement[]> => {
+    const userAchievements = await prisma.user_achievement.findMany({
+        where: { userId: user.userID },
+        include: { template: true },
+    });
+    const userAchievementGroups: { [group: string]: User_achievement[] } = {};
+    userAchievements.forEach((ua) => {
+        if (!userAchievementGroups[ua.template.group]) {
+            userAchievementGroups[ua.template.group] = [];
+        }
+        userAchievementGroups[ua.template.group].push(ua);
+    });
+    const achievements: Achievement[] = await generateReorderedAchievementData(userAchievementGroups, user);
+    return achievements;
+};
+
+const generateReorderedAchievementData = async (groups: { [group: string]: User_achievement[] }, user: User): Promise<Achievement[]> => {
+    const achievements: Achievement[] = [];
+    for (const group in groups) {
+        const sortedGroupAchievements = groups[group].sort((a, b) => a.groupOrder - b.groupOrder);
+        const groupAchievement: Achievement = await assembleAchievementData(sortedGroupAchievements, user);
+        achievements.push(groupAchievement);
+    }
+    return achievements;
+};
+
+const assembleAchievementData = async (userAchievements: User_achievement[], user: User): Promise<Achievement> => {
+    let currentAchievementIndex = userAchievements.findIndex((ua) => !ua.achievedAt);
+    currentAchievementIndex = currentAchievementIndex >= 0 ? currentAchievementIndex : userAchievements.length - 1;
+    const contextKeys = Object.keys(userAchievements[currentAchievementIndex].context);
+    const achievementContext: Partial<Context> = {
+        user: { ...user, fullName: `${user.firstname} ${user.lastname}` },
+    };
+    const newContextValue = await Promise.all(
+        contextKeys.map(async (key) => {
+            const [type, id] = userAchievements[currentAchievementIndex].context[key].split('/');
```

Comment:
I think this will not work, because we store not the relationId `match/1` in the user achievement context, we store other data like 
```
{
    matchId?: number
    subcourseId?: number
    matchPartner?: string
    ....
}
```